types.go

```diff
@@ -0,0 +1,108 @@
+// TODO(pav-kv): consider having type aliases for term and index.
+type entryID struct {
+	term  uint64
+	index uint64
+}
+
+// pbEntryID returns the ID of the given pb.Entry.
+func pbEntryID(entry *pb.Entry) entryID {
+	return entryID{term: entry.Term, index: entry.Index}
+}
+
+// logSlice describes a correct slice of a raft log.
+//
+// Every log slice is considered in a context of a specific leader term. This
+// term does not necessarily match entryID.term of the entries, since a leader
+// log contains both entries from its own term, and some earlier terms.
+//
+// Two slices with a matching logSlice.term are guaranteed to be consistent,
+// i.e. they never contain two different entries at the same index. The reverse
+// is not true: two slices with different logSlice.term may contain both
+// matching and mismatching entries. Specifically, logs at two different leader
+// terms share a common prefix, after which they *permanently* diverge.
+//
+// A well-formed logSlice conforms to raft safety properties. It provides the
+// following guarantees:
+//
+//  1. entries[i].Index == prev.index + 1 + i,
+//  2. prev.term <= entries[0].Term,
+//  3. entries[i-1].Term <= entries[i].Term,
+//  4. entries[len-1].Term <= term.
+//
+// Property (1) means the slice is contiguous. Properties (2) and (3) mean that
+// the terms of the entries in a log never regress. Property (4) means that a
+// leader log at a specific term never has entries from higher terms.
+//
+// Users of this struct can assume the invariants hold true. Exception is the
+// "gateway" code that initially constructs logSlice, such as when its content
+// is sourced from a message that was received via transport, or from Storage,
+// or in a test code that manually hard-codes this struct. In these cases, the
+// invariants should be validated using the valid() method.
+type logSlice struct {
+	// term is the leader term containing the given entries in its log.
+	term uint64
+	// prev is the ID of the entry immediately preceding the entries.
+	prev entryID
+	// entries contains the consecutive entries representing this slice.
+	entries []pb.Entry
+}
+
+// lastIndex returns the index of the last entry in this log slice. Returns
+// prev.index if there are no entries.
+func (a logSlice) lastIndex() uint64 {
+	return a.prev.index + uint64(len(a.entries))
+}
+
+// lastEntryID returns the ID of the last entry in this log slice, or prev if
+// there are no entries.
+func (a logSlice) lastEntryID() entryID {
+	if ln := len(a.entries); ln != 0 {
+		return pbEntryID(&a.entries[ln-1])
+	}
+	return a.prev
+}
+
+// valid returns nil iff the logSlice is a well-formed log slice. See logSlice
+// comment for details on what constitutes a valid raft log slice.
+func (a logSlice) valid() error {
+	prev := a.prev
+	for i := range a.entries {
+		id := pbEntryID(&a.entries[i])
+		if id.term < prev.term || id.index != prev.index+1 {
+			return fmt.Errorf("entries %+v and %+v not consistent", prev, id)
+		}
```

Comment:
```suggestion
		if id.term < prev.term || id.term > a.term || id.index != prev.index+1 {
			return fmt.Errorf("entries %+v and %+v not consistent, leader term: %d", prev, id, a.term)
		}
```

Reply:
We don't need to compare every entry `term` with `a.term`. The last entry has the biggest term, so it's enough to only check the last one (see a couple of lines below).

Reply:
Modified the error message to contain the leader term though, this would be useful for debugging.